File: scala/src/main/scala/task1.scala
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.json4s.jackson.JsonMethods._
import org.json4s._
import org.json4s.jackson.Serialization

import java.io.{File, PrintWriter}

object task1 {
    def main(args: Array[String]): Unit = {
        val conf: SparkConf = new SparkConf().setAppName("Task1").setMaster("local[*]")
        val sc:SparkContext = new SparkContext(conf)
        sc.setLogLevel("ERROR")
        val inputFilePath = args(0)
        val outputFilePath = args(1)

        val extracted_uid_bid_rdd = sc.textFile(inputFilePath)
            .map((row) => parse(row))
            .map((record) => {
                implicit val formats = DefaultFormats
                ((record \ "user_id").extract[String], (record \ "business_id").extract[String])
            })
        val uid_indexed_rdd = extracted_uid_bid_rdd.map((record) => record._1).distinct().zipWithIndex()
        val uid_to_index_map = uid_indexed_rdd.collectAsMap()
        val bid_indexed_rdd = extracted_uid_bid_rdd.map((record) => record._2).distinct().zipWithIndex()
        val bid_to_index_map = bid_indexed_rdd.collectAsMap()
        val index_to_bid_map = bid_to_index_map.map(_.swap)

        val bid_to_uid_map = extracted_uid_bid_rdd.map(tup => (bid_to_index_map.getOrElse(tup._2, throw new RuntimeException("test2")), uid_to_index_map.getOrElse(tup._1, throw new RuntimeException("test"))))
            .groupByKey()
            .map((record) => {
                (record._1, record._2.toSet)
            })
            .collectAsMap()
        val N = 70
        val buckets = uid_to_index_map.size

        val a_values = List.tabulate(N)(i => {
            val r1 = scala.util.Random
            r1.nextInt(911382371)
        })
        val b_values = List.tabulate(N)(i => {
            val r2 = scala.util.Random
            r2.nextInt(972667571)
        })
        val uid_hashed_indexed_rdd = uid_indexed_rdd.map((record) => {
            val r = scala.util.Random
            val hash = List.tabulate(N)(i => (((a_values(i) * record._2 + b_values(i)) % 1000000009) % buckets))
            (uid_to_index_map.getOrElse(record._1, 0L), hash)
        })
        val uid_aggr_bid_data = extracted_uid_bid_rdd
            .map(record => (uid_to_index_map.getOrElse(record._1, 0L), bid_to_index_map.getOrElse(record._2, 0L)))
            .groupByKey()
            .map((record) => {
                (record._1, record._2.toSet.toList)
            })
        def chooseLeastHashValue(list1: List[Long], list2: List[Long]) = {
            val list3 = for ((val1, val2) <- (list1 zip list2)) yield Math.min(val1, val2)
            list3
        }
        val sig_matrix_rdd = uid_aggr_bid_data.join(uid_hashed_indexed_rdd)
            .map(record => record._2)
            .flatMap(tup => {
                val n = tup._1.size
                List.tabulate(n)(i => (tup._1(i), tup._2))
            })
            .reduceByKey(chooseLeastHashValue).collect()

        val numBands = 70
        val bkt = collection.mutable.Map[Int, scala.collection.immutable.Set[Long]]()
        val candidates = collection.mutable.Set[List[Long]]()
        for (band_num  <- 0 until numBands - 1) {
            bkt.clear()
            val tup_hashCode_uid = sig_matrix_rdd.map((tup_uid_list_hash) => {
                val uid = tup_uid_list_hash._1
                val hashArr = tup_uid_list_hash._2
                val hashCode = hashArr(band_num)
                (hashCode, uid)
            })
            val set_of_uids = tup_hashCode_uid.groupBy(_._1).mapValues(_.map(_._2).toSet)
            for (uids <- set_of_uids.values) {
                if (uids.size > 1) {
                    for (uid_pair <- uids.subsets(2).map(_.toList)) {
                        candidates += uid_pair.sorted
                    }
                }
            }
        }
        def jaccardSim(set_1: Set[Long], set_2: Set[Long]) = {
            (set_1.intersect(set_2)).size / (set_1.union(set_2)).size.toFloat
        }
        val writer = new PrintWriter(new File(outputFilePath))
        for (candidate <- candidates) {
            val c1 = candidate(0)
            val c2 = candidate(1)

                val p1 = bid_to_uid_map(c1)
                val p2 = bid_to_uid_map(c2)
                val sim = jaccardSim(p1, p2)
                if (sim >= 0.05) {
                    implicit val formats = DefaultFormats
                    writer.write(Serialization.write(Map("b1" -> index_to_bid_map.getOrElse(c1, ""), "b2" -> index_to_bid_map.getOrElse(c2, ""), "sim" -> sim)) + "\n")
                }
        }
        writer.close()
    }
}

File: python/task1.py
from collections import defaultdict
from pyspark import SparkContext
import sys
from math import ceil, floor
from pyspark import SparkConf
from itertools import islice, combinations, chain
import time
import json
import random

class task1:
    
    def make_function(self, a, b, m):
        def hash_func(x_input):
            return ((a * x_input + b) % 1000000009) % m
        return hash_func

    def processInput(self, sc, input_file_path, output_file_path):
        extracted_uid_bid_rdd = sc.textFile(input_file_path) \
            .map(lambda row: json.loads(row)) \
            .map(lambda record: (record['user_id'], record['business_id'])).cache()
        
        user_id_indexed_rdd = extracted_uid_bid_rdd \
            .map(lambda record: record[0]) \
            .distinct() \
            .zipWithIndex()
        
        user_id_to_index_map = user_id_indexed_rdd.collectAsMap()

        business_id_indexed_rdd = extracted_uid_bid_rdd \
            .map(lambda record: record[1]) \
            .distinct() \
            .zipWithIndex()
        
        business_id_to_index_map = business_id_indexed_rdd.collectAsMap()
        index_to_bid_map = {v: k for k, v in business_id_to_index_map.items()}

        bid_to_uid_map = extracted_uid_bid_rdd.map(lambda tup: (business_id_to_index_map[tup[1]], user_id_to_index_map[tup[0]])) \
            .groupByKey() \
            .map(lambda tup: (tup[0], list(set(tup[1])))) \
            .collectAsMap()

        N = 70
        buckets = len(user_id_to_index_map)
        # Create Hash Functions
        hash_fns = []
        for i in range(N):
            a = random.randrange(1, 911382371, pow(10, 7) + 100 + 10 + 1)
            b = random.randrange(1, 972667571, pow(10, 7) + 100 + 10 + 1)
            hash_fns.append(self.make_function(a, b, buckets))
        
        user_hashed_indexes_rdd = user_id_indexed_rdd \
            .map(lambda uid_ind_tuple: (user_id_to_index_map[uid_ind_tuple[0]], \
                [func(uid_ind_tuple[1]) for func in hash_fns]))
        
        uid_aggr_bid_data = extracted_uid_bid_rdd \
            .map(lambda record: (user_id_to_index_map[record[0]], business_id_to_index_map[record[1]])) \
            .groupByKey() \
            .map(lambda record: (record[0], list(set(record[1]))))
        
        def chooseLeastHashValue(list1, list2):
            return [min(val1, val2) for (val1, val2) in zip(list1, list2)]
        
        # 1. Using Inner Join below
        sig_matrix_rdd = uid_aggr_bid_data.join(user_hashed_indexes_rdd) \
            .map(lambda tup: tup[1])\
            .flatMap(lambda tup: [(bid, tup[1]) for bid in tup[0]]) \
            .reduceByKey(chooseLeastHashValue)
        
        def createBands(hash_list, num_of_bands):
            N = len(hash_list) 
            chunk_lists = []
            number_of_rows = int(ceil(N / num_of_bands))
            for index, start in enumerate(range(0, N, number_of_rows)):
                chunk_lists.append((index, hash(tuple(hash_list[start: start + number_of_rows]))))
            return chunk_lists
        
        num_of_bands = 70
        candidates = sig_matrix_rdd \
            .flatMap(lambda tup: [(chunk, tup[0]) for chunk in createBands(tup[1], num_of_bands)]) \
            .groupByKey() \
            .map(lambda tup: tup[1]) \
            .filter(lambda businesses: len(businesses) > 1) \
            .flatMap(lambda businesses: [business_pair for business_pair in combinations(businesses, 2)])\
            .collect()
        
        def jaccardSim(s1, s2):
            s1 = set(s1)
            s2 = set(s2)
            return len(s1 & s2) / len(s1 | s2)
        
        res = []
        memo = set()
        for candidate in candidates:
            c1 = candidate[0]
            c2 = candidate[1]
            if ((c1, c2) not in memo and (c2, c1) not in memo):
                memo.add((c1, c2))
                sim = jaccardSim(bid_to_uid_map[c1], bid_to_uid_map[c2])
                if sim >= 0.05:
                    res.append({
                        "b1": index_to_bid_map[c1],
                        "b2": index_to_bid_map[c2],
                        "sim": sim
                    })
        with open(output_file_path, 'w+') as of:
            for obj in res:
                of.write(json.dumps(obj) + "\n")

        of.close()


    def main(self):
      input_file_path = sys.argv[1]
      output_file_path = sys.argv[2]
      conf = SparkConf().setAppName("task1").setMaster("local[*]")
      sc = SparkContext(conf=conf).getOrCreate()
      sc.setLogLevel("OFF")
      self.processInput(sc, input_file_path, output_file_path)



if __name__ == "__main__":
    t1 = task1()
    t1.main()